src/control/lib/control/telemetry.go

```diff
@@ -0,0 +1,357 @@
+	MetricsQueryReq struct {
+		Host        string   // host to query for telemetry data
+		Port        uint32   // port to use for collecting telemetry data
+		MetricNames []string // if empty, collects all metrics
+	}
+
+	// MetricsQueryResp contains the list of telemetry values per host.
+	MetricsQueryResp struct {
+		MetricSets []*MetricSet `json:"metric_sets"`
+	}
+)
+
+// MetricsQuery fetches the requested metrics values from the DAOS nodes.
+func MetricsQuery(ctx context.Context, req *MetricsQueryReq) (*MetricsQueryResp, error) {
+	if req == nil {
+		return nil, errors.New("nil request")
+	}
+
+	if req.Host == "" {
+		return nil, errors.New("host must be specified")
+	}
+
+	if req.Port == 0 {
+		return nil, errors.New("port must be specified")
+	}
+
+	scraped, err := scrapeMetrics(ctx, req.Host, req.Port)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to query metrics")
+	}
+
+	metricNames := req.MetricNames
+	if len(metricNames) == 0 {
+		// when caller doesn't specify metrics, we include all of them
+		metricNames = scraped.Keys()
+	}
+
+	return newMetricsQueryResp(scraped, metricNames)
+}
+
+func newMetricsQueryResp(scraped pbMetricMap, metricNames []string) (*MetricsQueryResp, error) {
+	resp := new(MetricsQueryResp)
+
+	list := make([]*MetricSet, 0)
+	for _, name := range metricNames {
+		mf, found := scraped[name]
+		if !found {
+			return nil, errors.Errorf("metric %q not found on host", name)
+		}
+
+		newSet := &MetricSet{
+			Name:        name,
+			Description: mf.GetHelp(),
+			Type:        metricTypeFromPrometheus(mf.GetType()),
+		}
+
+		for _, m := range mf.Metric {
+			newSet.Metrics = append(newSet.Metrics, getMetricFromPrometheus(m, mf.GetType()))
+		}
+
+		list = append(list, newSet)
+	}
+
+	resp.MetricSets = list
+	return resp, nil
+}
+
+func getMetricFromPrometheus(pMetric *pclient.Metric, metricType pclient.MetricType) Metric {
+	labels := metricsLabelsToMap(pMetric)
+	switch metricType {
+	case pclient.MetricType_COUNTER:
+		return newSimpleMetric(labels, pMetric.GetCounter().GetValue())
+	case pclient.MetricType_GAUGE:
+		return newSimpleMetric(labels, pMetric.GetGauge().GetValue())
+	case pclient.MetricType_SUMMARY:
+		summary := pMetric.GetSummary()
+		newMetric := &SummaryMetric{
+			Labels:      labels,
+			SampleSum:   summary.GetSampleSum(),
+			SampleCount: summary.GetSampleCount(),
+			Quantiles:   QuantileMap{},
+		}
+		for _, q := range summary.Quantile {
+			newMetric.Quantiles[q.GetQuantile()] = q.GetValue()
+		}
+		return newMetric
+	case pclient.MetricType_HISTOGRAM:
+		histogram := pMetric.GetHistogram()
+		newMetric := &HistogramMetric{
+			Labels:      labels,
+			SampleSum:   histogram.GetSampleSum(),
+			SampleCount: histogram.GetSampleCount(),
+		}
+		for _, b := range histogram.Bucket {
+			newMetric.Buckets = append(newMetric.Buckets,
+				&MetricBucket{
+					UpperBound:      b.GetUpperBound(),
+					CumulativeCount: b.GetCumulativeCount(),
+				})
+		}
+		return newMetric
+	}
```

Comment:
do we want to default to cover any unexpected types, or Will the default operation be to GeUntyped and that never fail ?

Reply:
I went back and forth on this myself. The more I think about it, the more I think we should deal with unknown types. My instinct is to skip anything we don't know how to deal with.

Reply:
Actually right now the Prometheus library I'm using to parse the endpoint data returns an error if there's an unknown metric type, so it shouldn't be possible at this point. I'm going to add a little logic to deal with it anyway in case that behavior changes.